pkg/config/redis.go

```diff
@@ -17,40 +17,48 @@ import (
 
 // Redis defines Redis client configuration.
 type Redis struct {
-	Address  string              `yaml:"address"`
-	Password string              `yaml:"password"`
+	Address  string `yaml:"address"`
+	Password string `yaml:"password"`
+	TLS      `yaml:",inline"`
 	Options  icingaredis.Options `yaml:"options"`
 }
 
 // NewClient prepares Redis client configuration,
 // calls redis.NewClient, but returns *icingaredis.Client.
 func (r *Redis) NewClient(logger *zap.SugaredLogger) (*icingaredis.Client, error) {
+	tlsConfig, err := r.TLS.MakeConfig(r.Address)
+	if err != nil {
+		return nil, err
+	}
+
 	c := redis.NewClient(&redis.Options{
 		Addr:        r.Address,
-		Dialer:      dialWithLogging(logger),
 		Password:    r.Password,
 		DB:          0, // Use default DB,
 		ReadTimeout: r.Options.Timeout,
+		TLSConfig:   tlsConfig,
 	})
 
 	opts := c.Options()
 	opts.MaxRetries = opts.PoolSize + 1 // https://github.com/go-redis/redis/issues/1737
+	opts.Dialer = dialWithLogging(opts.Dialer, logger)
 	c = redis.NewClient(opts)
```

Comment:
[When suggesting to just use `Options.TLSConfig`](https://github.com/Icinga/icingadb/pull/236#discussion_r706095468), I wasn't aware that go-redis just ignores that if you set a custom dialer. When [looking at the implementation](https://github.com/go-redis/redis/blob/fd3025bfed543a468085e18a9220b41632bf4f1d/options.go#L130-L141) it looks like the change from wrapping a plain `net.Dialer` to wrapping the one set up by `Option.init()` introduces some subtle changes:
1. It reduces the `DialTimeout` to 5 seconds (that change probably doesn't matter, it just shifts from the OS sending multiple TCP SYNs on the same socket to creating a new socket more often)
2. It increases `KeepAlive` from the default of ["currently 15 seconds"](https://pkg.go.dev/net#Dialer) to 5 minutes (not sure about that one, 5 minutes of inactivity could be long enough for some stateful firewall to drop the connection)

So the more I look at all this `redis.Options` defaults handling, the more I think we should rely on at little as possible (also that "create a new client and throw it away just to observe the side-effects of the unexported function `Options.init()`" feels quite hacky).

I think we haven't put any consideration into stuff like TCP keep-alives and that the Go default might be more suitable than the one of go-redis is sheer luck, but maybe just setting a custom dialer and doing all the things (including TLS) ourselves could be the better way.

Reply:
> doing all the things (including TLS) ourselves

That's exactly what I want to avoid.

Reply:
> > doing all the things (including TLS) ourselves
> 
> That's exactly what I want to avoid.

That's about 3 additional lines of code. So that shouldn't be a problem in my opinion.

> 1. It reduces the `DialTimeout` to 5 seconds (that change probably doesn't matter, it just shifts from the OS sending multiple TCP SYNs on the same socket to creating a new socket more often)

I also think it's just fine to not use a timeout here.

> 2. It increases `KeepAlive` from the default of ["currently 15 seconds"](https://pkg.go.dev/net#Dialer) to 5 minutes (not sure about that one, 5 minutes of inactivity could be long enough for some stateful firewall to drop the connection)

If this resulted in a retryable error, we would have no problem. But I would just stick to the default Go/OS mechanics.

Reply:
So all in all, I would put the TLS handling in our custom dialer too.

Reply:
> > 1. It increases `KeepAlive` from the default of ["currently 15 seconds"](https://pkg.go.dev/net#Dialer) to 5 minutes (not sure about that one, 5 minutes of inactivity could be long enough for some stateful firewall to drop the connection)
> 
> If this resulted in a retryable error, we would have no problem. But I would just stick to the default Go/OS mechanics.

Worst case that could theoretically happen is that a connection is idle (for example due to a XREAD blocking indefinitely), a stateful firewall silently dropping the connection and Icinga DB only noticing when it sends the next keepalive. Not sure how likely this is to happen, for example [WireGuard recommends](https://www.wireguard.com/quickstart/) a keepalive interval of 25 seconds, but not sure if that's a good reference as they're doing UDP.

Reply:
Shall we really re-invent that wheel and throw away the lib's (present and future) sane defaults?

Reply:
IMHO it's questionable if 5 minutes TCP keepalive is a sane default (at least for our project).